prototypes/ScheduledMerges.hs

```diff
@@ -1158,9 +1175,44 @@ depositNominalCredit (NominalDebt nominalDebt)
 -- Updates
 --
 
+-- | At some point, we want to merge the union level with the regular levels.
+-- We achieve this by moving it into a new last regular level once it is both
+-- completed (merged down to a single run) and fits into such a new level.
+--
+-- Our representation doesn't allow for empty levels, so we can only put the
+-- run directly after the pre-existing regular levels. If it is too large for
+-- that, we don't want to move it yet to avoid violating run size invariants
+-- and doing inefficient merges of runs with very different sizes.
+migrateUnionLevel :: forall s. Tracer (ST s) Event
+                  -> Counter -> Levels s -> UnionLevel s
+                  -> ST s (Levels s, UnionLevel s)
+migrateUnionLevel _ _ ls NoUnion = do
+    -- nothing to do
+    return (ls, NoUnion)
+migrateUnionLevel _tr _sc ls ul@(Union t _) =
+    -- TODO: tracing
+    getCompletedMergingTree t <&> \case
+      Just r
+        | null r ->
+          -- If the union level is empty, we can just drop it.
+          (ls, NoUnion)
+        | levellingRunSizeToLevel r <= length ls + 1 ->
+          -- If it fits into a hypothetical new last level, put it there.
+          --
+          -- TODO: In some cases it seems desirable to even add it to the
+          -- existing last regular level (so it becomes part of a merge
+          -- sooner), but that would lead to additional merging work that was
+          -- not accounted for. We'd need to be careful to ensure the merge
+          -- completes in time, without doing a lot of work in a short time.
+          (ls ++ [Level (Single r) []], NoUnion)
```

Comment:
A potential problem here is that we could build up too many levels. If all we do is union tables, and never perform any regular updates, then there is no guarantee that the finished union levels ever get merged together. We'd just add new levels indefinitely.

Reply:
I'm not sure what you mean. When creating a `union`, initially everything in there is part of the union level, where it will already be merged together. One the resulting run is moved to the regular levels, that table will never have a union level again.

Reply:
Hmm yeah maybe I just got confused, makes sense